pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -1039,10 +1039,395 @@ func TestValidatePipelineWorkspaces_Failure(t *testing.T) {
+				}},
+				Finally: []PipelineTask{{
+					Name: "final-task",
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "undefined parameter variable in final task",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Params: []ParamSpec{{
+					Name: "foo", Type: ParamTypeString,
+				}},
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+					Params: []Param{{
+						Name: "final-param", Value: ArrayOrString{Type: ParamTypeString, StringVal: "$(params.foo) and $(params.does-not-exist)"},
+					}},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline - final task with duplicate resource declaration",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Resources: []PipelineDeclaredResource{{
+					Name: "duplicate-resource", Type: PipelineResourceTypeGit,
+				}, {
+					Name: "duplicate-resource", Type: PipelineResourceTypeGit,
+				}},
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "foo-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "foo-task"},
+					Resources: &PipelineTaskResources{
+						Inputs: []PipelineTaskInputResource{{
+							Name: "the-resource", Resource: "duplicate-resource",
+						}},
+					},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline with invalid final tasks",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:     "final-task-1",
+					TaskRef:  &TaskRef{Name: "final-task"},
+					RunAfter: []string{"non-final-task"},
+				}, {
+					Name:    "final-task-2",
+					TaskRef: &TaskRef{Name: "final-task"},
+					Conditions: []PipelineTaskCondition{{
+						ConditionRef: "some-condition",
+					}},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline - workspace bindings in final task relying on a non-existent pipeline workspace",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name: "non-final-task", TaskRef: &TaskRef{Name: "foo"},
+				}},
+				Finally: []PipelineTask{{
+					Name: "final-task", TaskRef: &TaskRef{Name: "foo"},
+					Workspaces: []WorkspacePipelineTaskBinding{{
+						Name:      "shared-workspace",
+						Workspace: "pipeline-shared-workspace",
+					}},
+				}},
+				Workspaces: []WorkspacePipelineDeclaration{{
+					Name: "foo",
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline with empty finally",
```

Comment:
is this the same as "invalid pipeline with empty finally section"?

Reply:
nope, this is validating, empty finally without any DAG tasks under tasks section, updated description for both